include/merlin/memory_pool.cuh

```diff
@@ -0,0 +1,600 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cuda_runtime_api.h>
+#include <algorithm>
+#include <array>
+#include <functional>
+#include <iostream>
+#include <memory>
+#include <mutex>
+#include <thread>
+#include <vector>
+#include "debug.hpp"
+
+/**
+ * Allocators are used by the memory pool (and maybe other classes) to create
+ * RAII complient containers for buffers allocated in different memory areas.
+ */
+template <class T, class Allocator>
+struct AllocatorBase {
+  using type = T;
+  using sync_unique_ptr = std::unique_ptr<type, Allocator>;
+  using async_unique_ptr = std::unique_ptr<type, std::function<void(type*)>>;
+  using shared_ptr = std::shared_ptr<type>;
+
+  inline static sync_unique_ptr make_unique(size_t n) {
+    return sync_unique_ptr(Allocator::alloc(n));
+  }
+
+  inline static async_unique_ptr make_unique(size_t n, cudaStream_t ctx) {
+    return {Allocator::alloc(n, ctx), [ctx](type* p) { Allocator::free(p); }};
+  }
+
+  inline static shared_ptr make_shared(size_t n, cudaStream_t ctx = nullptr) {
+    return {Allocator::alloc(n, ctx),
+            [ctx](type* p) { Allocator::free(p, ctx); }};
+  }
+
+  inline void operator()(type* ptr) { Allocator::free(ptr, nullptr); }
+};
+
+/**
+ * Trivial fallback implementation using the standard C++ allocator. This mostly
+ * exists to ensure interface correctness, and as an illustration of what a
+ * proper allocator implementation should look like.
+ */
+template <class T>
+struct StandardAllocator final : AllocatorBase<T, StandardAllocator<T>> {
+  using type = typename AllocatorBase<T, StandardAllocator<T>>::type;
+
+  static constexpr const char* name{"StandardAllocator"};
+
+  inline static type* alloc(size_t n, cudaStream_t ctx = nullptr) {
+    return new type[n];
+  }
+
+  inline static void free(type* ptr, cudaStream_t ctx = nullptr) {
```

Comment:
Usually, the cudaStream_t  has no nullptr, it's not a pointer, can be changed to `cudaStream_t stream = 0`.
Should be modified in other part of this PR.

Reply:
I am pretty sure it is internally implemented as a pointer to a struct, and not as an integer offset of in some resource array. Not sure if that is exposed externally in the runtime though.

Reply:
That `0` == `nullptr` == default stream is a different story.

Reply:
Well... let's cut this short. Do you want me to rename:
`ctx = nullptr` => `stream = 0` ?

Reply:
This is true and not at the same time. Technically speaking a `cudaStream_t` is a pointer to a struct (see below). So `nullptr` is the right C++-11 construct. CUDA predates C++-11, so `0` is used at most locations in the CUDA runtime because `NULL` is platform dependent. However, and that his where the numeric interpretation makes sense. Not only the value `0` (which coincidentally == nullptr) is has special meaning, also the values 0x1 (Legacy CUDA stream) and 0x2 (`thread_local` CUDA stream) gave special meaning. However, these values are almost never used nowadays. At least, I haven't seen a single application using those.

See also `driver_types.h` in the CUDA runtime:
```c++
/*******************************************************************************
*                                                                              *
*  SHORTHAND TYPE DEFINITION USED BY RUNTIME API                               *
*                                                                              *
*******************************************************************************/

/**
 * CUDA Error types
 */
typedef __device_builtin__ enum cudaError cudaError_t;

/**
 * CUDA stream
 */
typedef __device_builtin__ struct CUstream_st *cudaStream_t;

/**
 * CUDA event types
 */
typedef __device_builtin__ struct CUevent_st *cudaEvent_t;

/**
 * CUDA graphics resource types
 */
typedef __device_builtin__ struct cudaGraphicsResource *cudaGraphicsResource_t;
```

Reply:
Resolved as discussed in Slack!